synapse/handlers/federation.py

```diff
@@ -234,8 +234,16 @@ def log_failure(f):
 
         if event.type == EventTypes.Member:
             if event.membership == Membership.JOIN:
-                user = UserID.from_string(event.state_key)
-                yield user_joined_room(self.distributor, user, event.room_id)
+                context = yield self.state_handler.compute_event_context(
+                    event, old_state=state, outlier=event.internal_metadata.is_outlier()
+                )
```

Comment:
We already have the context as its returned by `_handle_new_event` (at line 180)